examples/pandas/split-apply-combine/my_functions.py

```diff
@@ -0,0 +1,179 @@
+            if np.isnan(x["tax_credit"]) else
+            f"({int(x['income'])} * {x['tax_rate']}) - ({int(x['income'])} * {x['tax_rate']}) * {x['tax_credit']}"
+        ), axis=1
+    )
+    return df["Tax Formula"]
+
+
+def tax(tax_formula: Series) -> Series:
+    """
+    Return a series 'Tax' from 'Tax Formula' series.
+    :param tax_formula: the 'Tax Formula' series.
+    :return: the 'Tax Formula' Series
+    """
+    df = tax_formula.to_frame()
+    df["Tax"] = df["Tax Formula"].apply(lambda x: round(pandas.eval(x)))
+    return df["Tax"]
+
+
+# ----------------------------------------------------------------------------------------------------------------------
+# Tax calculation pipelines (chained set of transformations)
+# ----------------------------------------------------------------------------------------------------------------------
+
+@pipe(
+    step(tax_rate, tax_rates=tax_rates_rules),  # apply the tax_rate step
+    step(tax_credit, tax_credits=tax_credits_rules),  # apply the tax_credit step
+)
+def under_100k_tax(under_100k: DataFrame) -> DataFrame:
+    """
+    Tax calculation pipeline for 'Income' under 100k.
+    :param df: The DataFrame  where 'Income' is under 100k
+    :return: the DataFrame with the 'Tax' Series
+    """
+    return under_100k
+
+
+@pipe(
+    step(tax_rate, tax_rates=tax_rates_rules),  # apply the tax_rate step
+)
+def over_100k_tax(over_100k: DataFrame) -> DataFrame:
+    """
+    Tax calculation pipeline for 'Income' over 100k.
+    :param over_100k: The DataFrame where 'Income' is over 100k
+    :return: the DataFrame with the 'Tax' Series
+    """
+    return over_100k
+
+
+# ----------------------------------------------------------------------------------------------------------------------
+# Tax calculator dataflow
+# ----------------------------------------------------------------------------------------------------------------------
+
+@extract_fields(
+    {"under_100k": DataFrame, "over_100k": DataFrame}
+)
+# This is the first node in the DAG
+def split_dataframe(input: DataFrame) -> Dict[str, DataFrame]:
+    """
+    That function takes the DataFrame in input and split it in 2 DataFrames:
+      - under_100k: Rows where 'Income' is under 100k
+      - over_100k: Rows where 'Income' is over 100k
+    """
+    return {
+        "under_100k": input.query('Income < 100000'),
+        "over_100k": input.query('Income >= 100000'),
+    }
+
+
+@extract_columns('Income', 'Tax Rate', 'Tax Credit')
+def combine_dataframe(under_100k_tax: DataFrame, over_100k_tax: DataFrame) -> DataFrame:
+    """
+    That function combine the DataFrames under_100k and over_100k
+
+    The @extract_columns decorator is making the Series available for processing.
+    """
+    combined = pd.concat([under_100k_tax, over_100k_tax], axis=0).sort_index()
+    return combined
+
+
+# This is the last node in the DAG
+def end(combine_dataframe: DataFrame, tax_formula: Series, tax: Series) -> DataFrame:
```

Comment:
We should name it like a variable / something more descriptive.
```suggestion
def final_tax_dataframe(combine_dataframe: DataFrame, tax_formula: Series, tax: Series) -> DataFrame:
```